Composer/packages/server/src/models/connector/csharpBotConnector.ts

```diff
@@ -2,74 +2,178 @@
-    this.adminEndpoint = adminEndpoint;
+  static botRuntimes: { [key: string]: ChildProcess } = {};
+  constructor(endpoint: string) {
     this.endpoint = endpoint;
   }
+  static stopAll = (signal: string) => {
+    for (const pid in CSharpBotConnector.botRuntimes) {
+      const runtime = CSharpBotConnector.botRuntimes[pid];
+      log(`kill runtime before exit at ${runtime.pid}`);
+      runtime.kill(signal);
+      delete CSharpBotConnector.botRuntimes[pid];
+    }
+  };
 
-  public status: BotStatus = BotStatus.NotConnected;
+  private stop = () => {
+    CSharpBotConnector.stopAll('SIGKILL');
+    this.status = BotStatus.NotConnected;
+  };
 
-  connect = async (_: BotEnvironments, __: string) => {
-    // confirm bot runtime is listening here
-    try {
-      await axios.get(this.adminEndpoint + '/api/admin');
-    } catch (err) {
-      throw new Error(err);
+  private buildProcess = async (dir: string): Promise<number | null> => {
+    return new Promise((resolve, reject) => {
+      const build = spawn(`pwsh ./Scripts/build_runtime.ps1`, {
+        cwd: dir,
+        detached: true,
+        shell: true,
+        stdio: ['ignore', 'ignore', 'inherit'],
+      });
+      log(`build pid : ${build.pid}`);
+
+      build.stderr &&
+        build.stderr.on('data', function(err) {
+          reject(err.toString());
+        });
+
+      build.on('exit', function(code) {
+        resolve(code);
+      });
+    });
+  };
+
+  private getConnectorConfig = (config: DialogSetting) => {
+    const configList: string[] = [];
+    if (config.MicrosoftAppPassword) {
+      configList.push('--MicrosoftAppPassword');
+      configList.push(config.MicrosoftAppPassword);
+    }
+    if (config.luis) {
+      if (config.luis.authoringKey) {
+        configList.push('--luis:endpointKey');
+        configList.push(config.luis.authoringKey);
+      }
+      if (config.luis.authoringRegion) {
+        configList.push('--luis:endpoint');
+        configList.push(`https://${config.luis.authoringRegion}.api.cognitive.microsoft.com`);
+      }
     }
 
-    this.status = BotStatus.NotConnected;
+    return configList;
+  };
+  private addListeners = (child: ChildProcess, handler: Function) => {
+    if (child.stdout !== null) {
+      child.stdout.on('data', (data: any) => {
+        log(`stdout: ${data}`);
+      });
+    }
+
+    if (child.stderr !== null) {
+      child.stderr.on('data', (data: any) => {
+        log(`stderr: ${data}`);
+      });
+    }
+
+    child.on('close', code => {
+      log(`close ${code}`);
+      handler();
+    });
+
+    child.on('error', (err: any) => {
+      log(`stderr: ${err}`);
+    });
+
+    child.on('exit', code => {
+      log(`exit: ${code}`);
+      handler();
+    });
 
-    return `${this.endpoint}/api/messages`;
+    child.on('message', msg => {
+      log(msg);
+    });
+
+    child.on('disconnect', code => {
+      log(`disconnect: ${code}`);
+      handler();
+    });
```

Comment:
Let's add the pid to this as well. Maybe prefix each log with the pid?

```ts
log('bot runtime (%d): some message %s', child.pid, data) 
```

Reply:
_**prefix each log with the pid**_ Do you means` log('%d ……', child.pid, …) `or extend a new debugger using pid as sub namespace? like `log.extend(child.pid)`